Run systemd/rootless when systemd/rootless unit testing

robmry opened the pull request:
**- What I did**

- https://github.com/moby/moby/pull/49659
  - Spotted the firewalld tests weren't enabling firewalld, and started fixing that
  - A unit test failed with firewalld enabled, which was a clue that systemd wasn't enabled for its own tests
  - Rootless was in the same position - unit tests in rootless mode don't seem useful, but that can be dealt with separately.

#### Run systemd/rootless when systemd/rootless unit testing

**- What Pawel did**

#### hack: Fix TestOverlay* test failure in pkg/archive
The pkg/archive/archive_linux_test.go tests create a test archive content in a temporary directory. This also includes device nodes (using mknod).

Running these tests in Docker-in-Docker (dind) with systemd was failing with "operation not permitted" because the rootfs is remounted with `nodev`.

This change aligns `hack/dind-systemd` with `hack/dind` by conditionally mounting a `tmpfs` on `/tmp` (with dev enabled), to make the mknod work.

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
```markdown changelog

```

robmry committed:
Run systemd/rootless when systemd/rootless unit testing

Signed-off-by: Rob Murray <rob.murray@docker.com>

vvoland committed:
hack: Fix TestOverlay* test failure in pkg/archive

The pkg/archive/archive_linux_test.go tests create a test archive
content in a temporary directory. This also includes device nodes (using
mknod).
Running these tests in Docker-in-Docker (dind) with systemd was failing
with "operation not permitted" because the rootfs is remounted with
`nodev`.

This change aligns `hack/dind-systemd` with `hack/dind` by
conditionally mounting a `tmpfs` on `/tmp` (with dev enabled),
to make the mknod work.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

AkihiroSuda commented on .github/workflows/.test.yml:
```diff
@@ -83,10 +83,17 @@ jobs:
         run: |
           CACHE_DEV_SCOPE=dev
+          if [[ "${{ matrix.mode }}" == *"rootless"* ]]; then
+            echo "DOCKER_ROOTLESS=1" >> $GITHUB_ENV
```
What does this mean for unit tests?

robmry replied:
The tests run in moby's dev container, and that env is used when the container's created.

robmry replied:
But I don't think there's a good reason to run unit tests in rootless mode - they'll just run directly in the dev container ... so, same as the other tests. I think it can be removed.

The more-important one was systemd, but it seemed misleading to call a test run "unit (rootless)" without enabling rootless.

robmry replied:
I've updated the description to explain how I got here. We can disable the rootless unit tests separately. (I'd not looked at any of this config before, and my main goal is to unblock the firewalld testing - trying to change the test matrix seemed a bit ambitious!)

AkihiroSuda replied:
LGTM, but could you duplicate that description to the code comment lines

robmry replied:
Done.

tianon commented on hack/dind-systemd:
```diff
@@ -19,6 +19,11 @@ if [ ! -t 0 ]; then
+# Mount tmpfs at /tmp with dev flag
+# We need the dev flag to allow mknod to work for the
+# pkg/archive/archive_linux_test.go tests.
+mount -t tmpfs -o rw,dev tmpfs /tmp
```
IMO this is abusing the `--privileged` this has to run with a little bit, because it's something we _could_ reasonably do to the container instead of having the container do itself -- what if we instead added `--tmpfs /tmp:dev` in here?

https://github.com/moby/moby/blob/e1e6d35277b8abe7e2173f98e729eb14569ab8f7/Makefile#L118

vvoland replied:
That would work too. `dind-systemd` made sense to me because it's reliant on the container being `--privileged` anyway (as the whole `dind` is). It also keeps all the systemd shenanigans in one place.

But I'm fine either way.

robmry replied:
The non-systemd `dind` script already conditionally mounts `/tmp` in the container ...

https://github.com/moby/moby/blob/c43aa0b6aa7c88343f0951ba9a39c69aa51c54ef/hack/dind#L54-L57

robmry replied:
@tianon - how do we unblock this / how strongly do you feel about moving the mount?

Keeping the systemd config together makes sense to me, there's already a mount of `/tmp` in the other script, and the container will always be privileged. There's more than one way to achieve the same thing, I'm not sure that means this way is wrong ... unless there are consequences that need to be avoided?

tianon replied:
Doh, I didn't remember the other script having it 😅  if we use the same implementation (with the conditional), I withdraw my complaint ❤️

(Honestly I wish all of this script were somehow possible with dockerd args, but alas, that's a much bigger set of changes 🙈)

robmry replied:
Done.

vvoland reviewed (approved):
half LGTM 😅

robmry commented:
Updated to address comments, rebased, and swapped the order of the commits (to fix the test before trying to run it).

Considered moving the "/tmp" mount to the Makefile as @tianon suggested ... because it's not actually a systemd-specific issue (it worked for non-systemd because the conditional mount was already in its hack/dind script). But, wasn't sure-enough that everybody runs the container via the Makefile, and didn't think it worth causing disruption by removing the conditional mount. Also considered checking "/tmp" is mounted "rw,exec,dev", but that can be left for a follow-up if needed.

robmry commented:
Some unrelated flakiness ...

```
=== FAIL: github.com/docker/docker/integration/container TestWindowsDevices/process/no_device_mounted (60.59s)
    devices_windows_test.go:95: assertion failed: error is not nil: error during connect: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.48/containers/create": EOF

=== FAIL: libnetwork/networkdb TestNetworkDBCRUDMediumCluster (22.01s)
    networkdb_test.go:429: timeout hit after 20s: node3:Waiting for cluster peers to be established

=== FAIL: amd64.integration.container TestStopContainerWithTimeout/unlimited_timeout:_expect_graceful_container_stop (0.28s)
    stop_test.go:108: assertion failed: 2 (inspect.State.ExitCode int) != 42 (tc.expectedExitCode int)
    --- FAIL: TestStopContainerWithTimeout/unlimited_timeout:_expect_graceful_container_stop (0.28s)

=== FAIL: amd64.integration.container TestStopContainerWithTimeout/big_enough_timeout:_expect_graceful_container_stop (0.30s)
    stop_test.go:108: assertion failed: 2 (inspect.State.ExitCode int) != 42 (tc.expectedExitCode int)
    --- FAIL: TestStopContainerWithTimeout/big_enough_timeout:_expect_graceful_container_stop (0.30s)
```

tianon reviewed (approved):
LGTM

As a point of discipline, we should avoid `{{ ... }}` entirely inside code blocks in GHA (passing them as env instead so they get escaped safely), but that's preexisting not new in this PR. 👍
